### PR TITLE
fix(pwt): compatibility in CWD with wrong casing

### LINKED ISSUE #9193: [BUG] test runner breaks when invoked by path name with the wrong casing
Note: this is ultimately a problem between chair and keyboard but I'm filing an issue as it might be worth considering to make Playwright more robust against the problem described here. If not, maybe it'll at least help someone else who runs into this in the future.

**Context:**
- Playwright Version: 1.15.0
- Operating System: Windows
- Node.js version: 16.10.0 (was also able to repro in 14.17.6 LTS)
- Browser: Chromium / All

**Code Snippet**

Repo: https://github.com/fkleuver/playwright-test-repro
(although it should be reproducible with any setup)

Simplest steps to repro:
```
cd C:\projects
git clone git@github.com:fkleuver/playwright-test-repro.git
cd playwright-test-repro
npm ci
npm run --prefix ../Playwright-test-repro e2e
                  # ^ note the capital P 
```

The above repro steps fail with the following error:
```
Error: test\w3.spec.js:3:1: test() can only be called in a test file
    at errorWithLocation (C:\projects\Playwright-test-repro\node_modules\@playwright\test\lib\test\util.js:195:10)
    at TestTypeImpl._createTest (C:\projects\Playwright-test-repro\node_modules\@playwright\test\lib\test\testType.js:78:51)
    at C:\projects\Playwright-test-repro\node_modules\@playwright\test\lib\test\transform.js:133:12
    at file:///C:/projects/Playwright-test-repro/test/w3.spec.js:3:1
    at ModuleJob.run (node:internal/modules/esm/module_job:183:25)
    at Loader.import (node:internal/modules/esm/loader:178:24)
    at importModuleDynamicallyWrapper (node:internal/vm/module:437:15)
    at Loader._requireOrImport (C:\projects\playwright-test-repro\node_modules\@playwright\test\lib\test\loader.js:214:52)
    at Loader.loadTestFile (C:\projects\playwright-test-repro\node_modules\@playwright\test\lib\test\loader.js:130:7)
    at Runner._run (C:\projects\playwright-test-repro\node_modules\@playwright\test\lib\test\runner.js:219:40)
```

The repro steps may seem a bit contrived but it's simply the easiest way to reproduce if you folks want to debug this.

**Describe the bug**

The "culprit" is module-scoped variables in Playwright which are used as a cache.
File: `src/test/globals.ts` (fns `setCurrentlyLoadingFileSuite` / `currentlyLoadingFileSuite`)

Apparently when the entry point is invoked with the wrong casing, two of those modules end up being loaded. `setCurrentlyLoadingFileSuite` is called on one instance of the module whereas `currentlyLoadingFileSuite` is called on another instance (where the backing `currentFileSuite` variable is always `undefined`)

How I ran into this: I had a shortcut to open `cmd.exe` in `C:\Projects\my-project` (the actual folder is `C:\projects\my-project`). Typo, my bad, but it worked fine until I switched from mocha to the playwright test runner.

Since the error itself doesn't really help (nor does googling for it), it took a while to find out the root cause.
Imagine the fun:
- run tests: fails
- try a bunch of things, up to eventually `cd .. && rmdir my-project /s /q` followed by a fresh git clone & install
- run tests: succeeds
- open another command prompt to the same folder, run tests: fails (?!)

I don't know the inner workings and constraints of Playwright well enough to confidently suggest any particular solution, but at least detecting duplicate module instantiation to throw a more helpful error in this scenario would seem like good form.

Something like this at the top of `globals.ts` would accomplish just that:
```ts
const s = Symbol.for('__playwright');
if (globalThis[s]) {
  throw new Error(`more than one instance of globals.js was loaded`);
} else {
  Reflect.defineProperty(globalThis, s, { value: true, configurable: true });
}
```

Then the error will directly point to this specific issue if it's caused by duplicate module instantiation, rather than potentially send the user on a trail to nowhere :-)

### LINKED ISSUE #16347: [BUG] Playwright breaks in case of current directory case mismatch
**Context:**
- Playwright Version: 1.24.2
- Browser: All

~~~
## System:
 - OS: Windows 10 10.0.19044
 - Memory: 2.98 GB / 15.93 GB
## Binaries:
 - Node: 16.14.0 - C:\Program Files\nodejs\node.EXE
 - Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
 - npm: 8.3.1 - C:\Program Files\nodejs\npm.CMD
## Languages:
 - Bash: 4.4.20 - C:\Windows\system32\bash.EXE
~~~

**Describe the bug**

If the current directory doesn't have "correct" uppercase/lowercase letters, playwright won't run any tests and gives a confusing error message because it ignores _playwright.config.ts_ and treats the tests themselves as configuration file(s). For example, it might say "Playwright Test did not expect test.beforeEach() to be called here".

**Code Snippet**

Powershell transcript using on [the example shipped provided with playwright](https://playwright.dev/docs/intro):

~~~
PS C:\Dev\Temp\playwright-test> npm init playwright@latest
...
PS C:\Dev\Temp\playwright-test> npx playwright test

Running 3 tests using 3 workers

  3 passed (4s)

To open last HTML report run:

  npx playwright show-report

PS C:\Dev\Temp\playwright-test> cd /dev/temp/playwright-test
PS C:\dev\temp\playwright-test> npx playwright test

Running 0 test using 0 worker


Error: ui-tests\example.spec.ts:5:16: Playwright Test did not expect test() to be called here.
Most common reasons include:
- You are calling test() in a configuration file.
- You are calling test() in a file that is imported by the configuration file.
- You have two different versions of @playwright/test. This usually happens
  when one of the dependencies in your package.json depends on @playwright/test.
...
~~~

The bug was originally described in #13092.

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes https://github.com/microsoft/playwright/issues/16347
Fixes https://github.com/microsoft/playwright/issues/9193

This bug only affects PowerShell 5.X and 6.X experience, when using PowerShell 7.X its not possible anymore to enter a directory with incorrect casing. https://github.com/PowerShell/PowerShell/pull/9250 was fixing it upstream - this patch workarounds it and normalizes paths before requiring them.

### CODE REVIEW

File: tests/installation/playwright.config.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ const config: PlaywrightTestConfig = {
   timeout: 5 * 60 * 1000,
   retries: 0,
   reporter: process.env.CI ? [
-    ['dot'],
+    ['list'],
```
Comment (mxschmitt) [inline tests/installation/playwright.config.ts:30]: we only have a small number of long running tests, so list is better there.

File: packages/playwright-test/src/loader.ts [hunk 2]
```diff
@@ -678,3 +679,19 @@ export function folderIsModule(folder: string): boolean {
   // Rely on `require` internal caching logic.
   return require(packageJsonPath).type === 'module';
 }
+
+function fixFilepathCapitalization(file: string): string {
+  /**
```
Comment (mxschmitt) [inline packages/playwright-test/src/loader.ts:684]: should we a 
```ts
if (process.platform !== 'win32')
   return;
```
there?

Comment (aslushnikov) [reply packages/playwright-test/src/loader.ts:684]: I'd call the method `fixWin32FilepathCapitalization` and call it on win32 platform only

File: packages/playwright-test/src/loader.ts [hunk 3]
```diff
@@ -678,3 +680,19 @@ export function folderIsModule(folder: string): boolean {
...
+
+function fixWin32FilepathCapitalization(file: string): string {
+  /**
+   * On Windows with PowerShell <= 6 it is possible to have a CWD with different
+   * casing than what the actual directory on the filesystem is. This can causes
+   * that we require the file multiple times with different casing. To mitigate
+   * this we get the actual underlying filesystem path and use that.
+   * https://github.com/microsoft/playwright/issues/9193#issuecomment-1219362150
+   */
+  const realFile = fs.realpathSync.native(file);
```
Comment (yury-s) [inline packages/playwright-test/src/loader.ts:692]: Does it have to be sync? It affects every _requireOrImport so might be noticeable on windows.

File: packages/playwright-test/src/loader.ts [hunk 4]
```diff
@@ -678,3 +680,19 @@ export function folderIsModule(folder: string): boolean {
   // Rely on `require` internal caching logic.
   return require(packageJsonPath).type === 'module';
 }
+
+function fixWin32FilepathCapitalization(file: string): string {
+  /**
+   * On Windows with PowerShell <= 6 it is possible to have a CWD with different
+   * casing than what the actual directory on the filesystem is. This can causes
```
Comment (yury-s) [inline packages/playwright-test/src/loader.ts:687]: typo: This can cause

**Comment by pavelfeldman** [comment]
I'm still fuzzy on what it going on here.

**Comment by mxschmitt** [comment]
Abstract problem: `@playwright\test\src\globals.ts` gets loaded two times:

- expectedcasing\node_modules\@playwright\test\src\globals.ts (2) (from pwt worker)
- eXpEcTeDcAsInG\node_modules\@playwright\test\src\globals.ts (1) (from test inside worker)

this ends up that the state inside is different, since these are different instances.

1) gets loaded `NPX -> pw-core CLI -> PWT CLI -> Worker-> globals.ts` - that this one has the `expected ` casing, which is good.

2) gets loaded from inside the test, since we load the test with an absolute path, which is different to what the 1) base path is in terms of casing, we end up loading the globals inside the runner two times.

e.g. `C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-incorrect-cwd-casing-installation-tests\eXpEcTeDcAsInG\sample.spec.js`

to it, so `@playwright\test\src\globals.ts` resolved into `eXpEcTeDcAsInG`.

---

1): setCurrentlyLoadingFileSuite

```
 itest:exec:stdout Error:
  itest:exec:stdout     at setCurrentlyLoadingFileSuite (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-
should-handle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\globals.ts:30:15)
  itest:exec:stdout     at Loader.loadTestFile (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-ha
ndle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\loader.ts:180:5)
  itest:exec:stdout     at Runner._run (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-inc
orrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\runner.ts:282:44)
  itest:exec:stdout     at TimeoutRunner.run (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-hand
le-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\playwright-core\lib\utils\timeoutRunner.js:53:14)
  itest:exec:stdout     at raceAgainstTimeout (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-han
dle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\playwright-core\lib\utils\timeoutRunner.js:113:15)
  itest:exec:stdout     at Runner.runAllTests (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-han
dle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\runner.ts:190:20)
  itest:exec:stdout     at runTests (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-incorr
ect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\cli.ts:163:18)
  itest:exec:stdout     at Gi.<anonymous> (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-
incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\cli.ts:65:7)
  itest:exec:stdout  +3s
  itest:exec:stdout requireOrImport: C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-incorr
ect-cwd-casing-installation-tests\eXpEcTeDcAsInG\sample.spec.js
  itest:exec:stdout  +0ms
```

2): currentlyLoadingFileSuite

```
  itest:exec:stdout Error:
  itest:exec:stdout     at currentlyLoadingFileSuite (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-sho
uld-handle-incorrect-cwd-casing-installation-tests\eXpEcTeDcAsInG\node_modules\@playwright\test\src\globals.ts:34:15)
  itest:exec:stdout     at TestTypeImpl._ensureCurrentSuite (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-t
est-should-handle-incorrect-cwd-casing-installation-tests\eXpEcTeDcAsInG\node_modules\@playwright\test\src\testType.ts:69:19)
  itest:exec:stdout     at TestTypeImpl._createTest (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-shou
ld-handle-incorrect-cwd-casing-installation-tests\eXpEcTeDcAsInG\node_modules\@playwright\test\src\testType.ts:85:24)
  itest:exec:stdout     at C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-handle-incorrect-cwd-ca
sing-installation-tests\eXpEcTeDcAsInG\node_modules\@playwright\test\src\transform.ts:244:12
  itest:exec:stdout     at Object.<anonymous> (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-han
dle-incorrect-cwd-casing-installation-tests\eXpEcTeDcAsInG\sample.spec.js:8:1)
  itest:exec:stdout     at Module._compile (node:internal/modules/cjs/loader:1103:14)
  itest:exec:stdout     at Module.A._compile (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-hand
le-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\lib\utilsBundleImpl.js:16:994)
  itest:exec:stdout     at Module._extensions..js (node:internal/modules/cjs/loader:1157:10)
  itest:exec:stdout     at Object.t.<computed>.tu._extensions.<computed> [as .js] (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-
cwd-ca-88ebc-ywright-test-should-handle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\lib\utilsBundleImpl.js:16:1010)
  itest:exec:stdout     at Module.load (node:internal/modules/cjs/loader:981:32)
  itest:exec:stdout     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
  itest:exec:stdout     at Module.require (node:internal/modules/cjs/loader:1005:19)
  itest:exec:stdout     at require (node:internal/modules/cjs/helpers:102:18)
  itest:exec:stdout     at Loader._requireOrImport (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-shoul
d-handle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\loader.ts:305:14)
  itest:exec:stdout     at Loader.loadTestFile (C:\Users\maxschmitt\AppData\Local\Temp\pwt\workspaces\playwright-test-should-handle-incorrect-cwd-ca-88ebc-ywright-test-should-ha
ndle-incorrect-cwd-casing-installation-tests\expectedcasing\node_modules\@playwright\test\src\loader.ts:182:18)
  itest:exec:stdout  +648ms
```

**Review by dgozman — APPROVED** [review]

**Pull request merged**